internal/statuspage/group_api_test.go

```diff
@@ -0,0 +1,233 @@
+package statuspage
+
+import (
+	"github.com/broadinstitute/revere/internal/statuspage/statuspagemocks"
+	"github.com/broadinstitute/revere/internal/statuspage/statuspagetypes"
+	"github.com/go-resty/resty/v2"
+	"github.com/jarcoal/httpmock"
+	"reflect"
+	"testing"
+)
+
+func TestDeleteGroup(t *testing.T) {
+	type args struct {
+		client  *resty.Client
+		pageID  string
+		groupID string
+	}
+	config := testConfig()
+	group := statuspagemocks.GroupFactory("to delete")
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "Succeeds on 204",
+			args: args{
+				client:  Client(config),
+				pageID:  config.Statuspage.PageID,
+				groupID: group.ID,
+			},
+		},
+		{
+			name: "Fails on 404",
+			args: args{
+				client:  Client(config),
+				pageID:  config.Statuspage.PageID,
+				groupID: "nonexistentID",
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
```

Comment:
I like the clarity of the test cases but the code blocks are hard to grok afterwards. 

If I were to add a new test case to check for a 401 or a specific api response, and I didn't get what I expected, I don't think i'd be able to easily debug. i.e. I'd like to be able to inspect the response, http code or even the headers to confirm expectations.

The blocks are also slightly different for each of the test blocks and it's not entirely clear why. e.g. TestGetGroups has stuff _after_ a `httpmock.DeactivateAndReset()`

Reply:
I agree. I wish I had more of a response. There's a _lot_ of boilerplate here. The fact that GoLand is happy to write most of it doesn't make it any easier to read after the fact. There's testing frameworks but at first glance they seem to break some of the IDE/tooling support, and what they seem best at abstracting is the "if x and y aren't equal, error and print why," which seems like the most readable part of the code that's here.

I think my hope was that by leaning on the codegen as much as possible the tests would at least look the same, as opposed to a bunch of edgecases like [these](https://github.com/broadinstitute/revere/blob/main/internal/shared/util_test.go). There's some variation still:
- [TestGetGroups](https://github.com/broadinstitute/revere/pull/9/files#diff-8548074129d8b56c20abe8b4eec9a06b5ee663914df1bb70317f91df89e64524R88-R105) and similar tests put some code after deactivate because we always need to deactivate the mock before returning--this pattern is used whenever a function that uses the mock can return either a value or an error: we don't know when we will return, so we can't wait until the end to deactivate the mock.
- [TestReconcileGroups](https://github.com/broadinstitute/revere/pull/9/files#diff-54bb83f9259976cdf4b9b6bedbf1508cfba5d63930df6ea52f78e1a7dbef3ee5R97-R112) has to check mutation side effects, so the boilerplate is subtly modified from the normal "got == want" pattern--this is mimicked for other mutation cases too.

Reply:
Re REST API handling/checking, Resty is fairly helpful about printing out legitimate errors it gets, and I just modified the CheckResponse function to include headers in addition to body and status code, that's a good catch, thank you. That stuff will show up in the test output if an error shows up unexpectedly, and an upside of Go's parsing is that unexpected response structure manifests as an error almost always.

Reply:
Thanks for explaining! So wait, does GoLand just spit out giant chunk of test code that *just work*? Thats both amazing and terrifying.

If I'm reading this correctly it looks like it the if statements are you tests/assertions/expectations and are testing:
a) an expected response error
b) an expected http response (or multiple for groups)
Is that right? I think It'd be helpful to add a comment before each if, considering how complex the test statements are and also so from a conceptual level is easy to see what parts of the requests are being tested.

Reply:
GoLand does indeed spit out huge chunks of code that work. Most of what I have to add is testing for side effects. It makes code testing inputs/outputs/errors, and if I need to do side-effecty things for HTTP mocks or whatnot around it then that's what I add. Good idea re comments, just added for both this file and the similar one for components.